google/cloud/pubsub/samples/samples.cc

```diff
@@ -1024,6 +1025,65 @@ void SubscriberRetrySettings(std::vector<std::string> const& argv) {
       std::move(p.second), __func__);
 }
 
+void OptimisticSubscribe(std::vector<std::string> const& argv) {
+  namespace examples = ::google::cloud::testing_util;
+  if (argv.size() != 3) {
+    throw examples::Usage{
+        "optimistic-subscribe <project-id> <topic-id> <subscription-id>"};
+  }
+  namespace pubsub_admin = ::google::cloud::pubsub_admin;
+  namespace pubsub = ::google::cloud::pubsub;
+  namespace gc = ::google::cloud;
+  [](std::string project_id, std::string topic_id,
+     std::string subscription_id) {
+    // Do not retry the attempts to consume messages.
+    auto subscriber = pubsub::Subscriber(
+        pubsub::MakeSubscriberConnection(
+            pubsub::Subscription(project_id, subscription_id)),
+        google::cloud::Options{}.set<pubsub::RetryPolicyOption>(
+            pubsub::LimitedErrorCountRetryPolicy(
+                /*maximum_failures=*/0)
+                .clone()));
+
+    // [START pubsub_optimistic_subscribe]
+    auto process_response = [](gc::StatusOr<pubsub::PullResponse> response) {
+      if (response) {
+        std::cout << "Received message " << response->message << "\n";
+        std::move(response->handler).ack();
+        return gc::Status();
+      }
+      if (response.status().code() == gc::StatusCode::kUnavailable &&
+          response.status().message() == "no messages returned") {
```

Comment:
hm, I don't love that we are recommending code that parses an error message. We make no guarantees about the contents of the error message. It's ok, though.

optional, and not for this PR: This could be an opportunity for that client origin stuff. Something that feels more official?